File: src/main/java/com/pyamsoft/padlock/dagger/PadLockDBImpl.java
/*
 * Copyright 2016 Peter Kenji Yamanaka
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.pyamsoft.padlock.dagger;

import android.content.Context;
import android.database.sqlite.SQLiteDatabase;
import android.database.sqlite.SQLiteOpenHelper;
import android.support.annotation.CheckResult;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.annotation.VisibleForTesting;
import android.text.TextUtils;
import com.pyamsoft.padlock.model.sql.PadLockEntry;
import com.pyamsoft.pydroidrx.SubscriptionHelper;
import com.squareup.sqlbrite.BriteDatabase;
import com.squareup.sqlbrite.SqlBrite;
import java.util.List;
import java.util.Locale;
import java.util.concurrent.TimeUnit;
import javax.inject.Inject;
import rx.Observable;
import rx.Scheduler;
import rx.Subscription;
import timber.log.Timber;

class PadLockDBImpl implements PadLockDB {

  @SuppressWarnings("WeakerAccess") @NonNull final BriteDatabase briteDatabase;
  @SuppressWarnings("WeakerAccess") @NonNull final PadLockOpenHelper openHelper;
  @NonNull private final Scheduler dbScheduler;
  @SuppressWarnings("WeakerAccess") @Nullable Subscription dbOpenSubscription;

  @Inject PadLockDBImpl(@NonNull Context context, @NonNull Scheduler scheduler) {
    dbScheduler = scheduler;
    openHelper = new PadLockOpenHelper(context);
    briteDatabase = new SqlBrite.Builder().build().wrapDatabaseHelper(openHelper, scheduler);
  }

  @SuppressWarnings("WeakerAccess") void openDatabase() {
    SubscriptionHelper.unsubscribe(dbOpenSubscription);

    // After a 1 minute timeout, close the DB
    dbOpenSubscription = Observable.timer(1, TimeUnit.MINUTES)
        .map(aLong -> {
          briteDatabase.close();
          return true;
        })
        .subscribeOn(dbScheduler)
        .observeOn(dbScheduler)
        .subscribe(ignoreMe -> Timber.d("PadLockDB is closed"),
            throwable -> Timber.e(throwable, "onError closing database"),
            () -> SubscriptionHelper.unsubscribe(dbOpenSubscription));
  }

  @Override @CheckResult @NonNull
  public Observable<Long> insert(@NonNull String packageName, @NonNull String activityName,
      @Nullable String lockCode, long lockUntilTime, long ignoreUntilTime, boolean isSystem,
      boolean whitelist) {
    final PadLockEntry entry =
        PadLockEntry.CREATOR.create(packageName, activityName, lockCode, lockUntilTime,
            ignoreUntilTime, isSystem, whitelist);
    if (PadLockEntry.isEmpty(entry)) {
      throw new RuntimeException("Cannot insert EMPTY entry");
    }

    return Observable.defer(() -> {
      Timber.i("DB: INSERT");
      openDatabase();
      return deleteWithPackageActivityNameUnguarded(packageName, activityName);
    }).map(deleted -> {
      Timber.d("Delete result: %d", deleted);
      return PadLockEntry.insertEntry(openHelper).executeProgram(entry);
    });
  }

  @NonNull @Override
  public Observable<Integer> updateIgnoreTime(long ignoreUntilTime, @NonNull String packageName,
      @NonNull String activityName) {
    if (packageName.equals(PadLockEntry.PACKAGE_EMPTY) || activityName.equals(
        PadLockEntry.ACTIVITY_EMPTY)) {
      throw new RuntimeException("Cannot update EMPTY entry");
    }

    return Observable.defer(() -> {
      Timber.i("DB: UPDATE IGNORE");
      openDatabase();
      return Observable.fromCallable(() -> PadLockEntry.updateIgnoreTime(openHelper)
          .executeProgram(ignoreUntilTime, packageName, activityName));
    });
  }

  @NonNull @Override
  public Observable<Integer> updateLockTime(long lockUntilTime, @NonNull String packageName,
      @NonNull String activityName) {
    if (packageName.equals(PadLockEntry.PACKAGE_EMPTY) || activityName.equals(
        PadLockEntry.ACTIVITY_EMPTY)) {
      throw new RuntimeException("Cannot update EMPTY entry");
    }

    return Observable.defer(() -> {
      Timber.i("DB: UPDATE LOCK");
      openDatabase();
      return Observable.fromCallable(() -> PadLockEntry.updateLockTime(openHelper)
          .executeProgram(lockUntilTime, packageName, activityName));
    });
  }

  @NonNull @Override
  public Observable<Integer> updateWhitelist(boolean whitelist, @NonNull String packageName,
      @NonNull String activityName) {
    if (packageName.equals(PadLockEntry.PACKAGE_EMPTY) || activityName.equals(
        PadLockEntry.ACTIVITY_EMPTY)) {
      throw new RuntimeException("Cannot update EMPTY entry");
    }

    return Observable.defer(() -> {
      Timber.i("DB: UPDATE WHITELIST");
      openDatabase();
      return Observable.fromCallable(() -> PadLockEntry.updateWhitelist(openHelper)
          .executeProgram(whitelist, packageName, activityName));
    });
  }

  @Override @NonNull @CheckResult
  public Observable<PadLockEntry.WithPackageActivityName> queryWithPackageActivityName(
      final @NonNull String packageName, final @NonNull String activityName) {
    return Observable.defer(() -> {
      Timber.i("DB: QUERY PACKAGE ACTIVITY");
      openDatabase();
      return briteDatabase.createQuery(PadLockEntry.TABLE_NAME,
          PadLockEntry.WITH_PACKAGE_ACTIVITY_NAME, packageName, activityName)
          .mapToOneOrDefault(PadLockEntry.WITH_PACKAGE_ACTIVITY_NAME_MAPPER::map,
              PadLockEntry.WithPackageActivityName.empty());
    });
  }

  /**
   * Get either the package with specific name of the PACKAGE entry
   *
   * SQLite only has bindings so we must make do
   * ?1 package name
   * ?2 the PadLock PACKAGE_TAG, see model.PadLockEntry
   * ?3 the specific activity name
   * ?4 the PadLock PACKAGE_TAG, see model.PadLockEntry
   * ?5 the specific activity name
   */
  @Override @NonNull @CheckResult
  public Observable<PadLockEntry> queryWithPackageActivityNameDefault(
      final @NonNull String packageName, final @NonNull String activityName) {
    return Observable.defer(() -> {
      Timber.i("DB: QUERY PACKAGE ACTIVITY DEFAULT");
      openDatabase();
      return briteDatabase.createQuery(PadLockEntry.TABLE_NAME,
          PadLockEntry.WITH_PACKAGE_ACTIVITY_NAME_DEFAULT, packageName,
          PadLockEntry.PACKAGE_ACTIVITY_NAME, activityName, PadLockEntry.PACKAGE_ACTIVITY_NAME,
          activityName)
          .mapToOneOrDefault(PadLockEntry.WITH_PACKAGE_ACTIVITY_NAME_DEFAULT_MAPPER::map,
              PadLockEntry.empty());
    });
  }

  @Override @NonNull @CheckResult
  public Observable<List<PadLockEntry.WithPackageName>> queryWithPackageName(
      final @NonNull String packageName) {
    return Observable.defer(() -> {
      Timber.i("DB: QUERY PACKAGE");
      openDatabase();
      return briteDatabase.createQuery(PadLockEntry.TABLE_NAME, PadLockEntry.WITH_PACKAGE_NAME,
          packageName).mapToList(PadLockEntry.WITH_PACKAGE_NAME_MAPPER::map);
    });
  }

  @Override @NonNull @CheckResult public Observable<List<PadLockEntry.AllEntries>> queryAll() {
    return Observable.defer(() -> {
      Timber.i("DB: QUERY ALL");
      openDatabase();
      return briteDatabase.createQuery(PadLockEntry.TABLE_NAME, PadLockEntry.ALL_ENTRIES)
          .mapToList(PadLockEntry.ALL_ENTRIES_MAPPER::map);
    });
  }

  @Override @NonNull @CheckResult
  public Observable<Integer> deleteWithPackageName(final @NonNull String packageName) {
    return Observable.defer(() -> {
      Timber.i("DB: DELETE PACKAGE");
      openDatabase();
      return Observable.fromCallable(
          () -> PadLockEntry.deletePackage(openHelper).executeProgram(packageName));
    });
  }

  @Override @NonNull @CheckResult
  public Observable<Integer> deleteWithPackageActivityName(final @NonNull String packageName,
      final @NonNull String activityName) {
    return Observable.defer(() -> {
      Timber.i("DB: DELETE PACKAGE ACTIVITY");
      openDatabase();
      return deleteWithPackageActivityNameUnguarded(packageName, activityName);
    });
  }

  @VisibleForTesting @NonNull @CheckResult
  Observable<Integer> deleteWithPackageActivityNameUnguarded(@NonNull String packageName,
      @NonNull String activityName) {
    return Observable.just(
        PadLockEntry.deletePackageActivity(openHelper).executeProgram(packageName, activityName));
  }

  @Override @NonNull @CheckResult public Observable<Integer> deleteAll() {
    return Observable.defer(() -> {
      Timber.i("DB: DELETE ALL");
      briteDatabase.execute(PadLockEntry.DELETE_ALL);
      SubscriptionHelper.unsubscribe(dbOpenSubscription);
      briteDatabase.close();
      deleteDatabase();
      return Observable.just(1);
    });
  }

  @Override public void deleteDatabase() {
    openHelper.deleteDatabase();
  }

  @SuppressWarnings("WeakerAccess") static class PadLockOpenHelper extends SQLiteOpenHelper {

    @NonNull private static final String DB_NAME = "padlock_db";
    private static final int DATABASE_VERSION = 4;
    @NonNull private final Context appContext;

    PadLockOpenHelper(final @NonNull Context context) {
      super(context.getApplicationContext(), DB_NAME, null, DATABASE_VERSION);
      appContext = context.getApplicationContext();
    }

    void deleteDatabase() {
      appContext.deleteDatabase(DB_NAME);
    }

    @Override public void onCreate(@NonNull SQLiteDatabase sqLiteDatabase) {
      Timber.d("onCreate");
      Timber.d("EXEC SQL: %s", PadLockEntry.CREATE_TABLE);
      sqLiteDatabase.execSQL(PadLockEntry.CREATE_TABLE);
    }

    @Override
    public void onUpgrade(@NonNull SQLiteDatabase sqLiteDatabase, int oldVersion, int newVersion) {
      Timber.d("onUpgrade from old version %d to new %d", oldVersion, newVersion);
      int currentVersion = oldVersion;
      if (currentVersion == 1 && newVersion >= 2) {
        upgradeVersion1To2(sqLiteDatabase);
        ++currentVersion;
      }

      if (currentVersion == 2 && newVersion >= 3) {
        upgradeVersion2To3(sqLiteDatabase);
        ++currentVersion;
      }

      if (currentVersion == 3 && newVersion >= 4) {
        upgradeVersion3To4(sqLiteDatabase);
        ++currentVersion;
      }
    }

    private void upgradeVersion3To4(SQLiteDatabase sqLiteDatabase) {
      Timber.d("Upgrading from Version 2 to 3 adds whitelist column");
      final String alterWithWhitelist = String.format(Locale.getDefault(),
          "ALTER TABLE %s ADD COLUMN %S INTEGER NOT NULL DEFAULT 0", PadLockEntry.TABLE_NAME,
          PadLockEntry.WHITELIST);

      Timber.d("EXEC SQL: %s", alterWithWhitelist);
      sqLiteDatabase.execSQL(alterWithWhitelist);
    }

    private void upgradeVersion2To3(SQLiteDatabase sqLiteDatabase) {
      Timber.d("Upgrading from Version 2 to 3 drops the whole table");

      final String dropOldTable =
          String.format(Locale.getDefault(), "DROP TABLE %s", PadLockEntry.TABLE_NAME);
      Timber.d("EXEC SQL: %s", dropOldTable);
      sqLiteDatabase.execSQL(dropOldTable);

      // Creating the table again
      onCreate(sqLiteDatabase);
    }

    private void upgradeVersion1To2(@NonNull SQLiteDatabase sqLiteDatabase) {
      Timber.d("Upgrading from Version 1 to 2 drops the displayName column");

      // Remove the columns we don't want anymore from the table's list of columns
      Timber.d("Gather a list of the remaining columns");
      final String[] updatedTableColumns = {
          PadLockEntry.PACKAGENAME, PadLockEntry.ACTIVITYNAME, PadLockEntry.LOCKCODE,
          PadLockEntry.LOCKUNTILTIME, PadLockEntry.IGNOREUNTILTIME, PadLockEntry.SYSTEMAPPLICATION
      };

      final String columnsSeperated = TextUtils.join(",", updatedTableColumns);
      Timber.d("Column seperated: %s", columnsSeperated);

      final String tableName = PadLockEntry.TABLE_NAME;
      final String oldTable = tableName + "_old";
      final String alterTable =
          String.format(Locale.getDefault(), "ALTER TABLE %s RENAME TO %s", tableName, oldTable);
      final String insertIntoNewTable =
          String.format(Locale.getDefault(), "INSERT INTO %s(%s) SELECT %s FROM %s", tableName,
              columnsSeperated, columnsSeperated, oldTable);
      final String dropOldTable = String.format(Locale.getDefault(), "DROP TABLE %s", oldTable);

      // Move the existing table to an old table
      Timber.d("EXEC SQL: %s", alterTable);
      sqLiteDatabase.execSQL(alterTable);

      onCreate(sqLiteDatabase);

      // Populating the table with the data
      Timber.d("EXEC SQL: %s", insertIntoNewTable);
      sqLiteDatabase.execSQL(insertIntoNewTable);

      Timber.d("EXEC SQL: %s", dropOldTable);
      sqLiteDatabase.execSQL(dropOldTable);
    }
  }
}

File: padlock-base/src/main/java/com/pyamsoft/padlock/base/PadLockDBImpl.kt
/*
 *     Copyright (C) 2017 Peter Kenji Yamanaka
 *
 *     This program is free software; you can redistribute it and/or modify
 *     it under the terms of the GNU General Public License as published by
 *     the Free Software Foundation; either version 2 of the License, or
 *     (at your option) any later version.
 *
 *     This program is distributed in the hope that it will be useful,
 *     but WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *     GNU General Public License for more details.
 *
 *     You should have received a copy of the GNU General Public License along
 *     with this program; if not, write to the Free Software Foundation, Inc.,
 *     51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */

package com.pyamsoft.padlock.base

import android.content.Context
import android.database.sqlite.SQLiteDatabase
import android.database.sqlite.SQLiteOpenHelper
import android.support.annotation.CheckResult
import android.text.TextUtils
import com.pyamsoft.padlock.api.PadLockDBDelete
import com.pyamsoft.padlock.api.PadLockDBInsert
import com.pyamsoft.padlock.api.PadLockDBQuery
import com.pyamsoft.padlock.api.PadLockDBUpdate
import com.pyamsoft.padlock.model.PadLockEntry
import com.pyamsoft.padlock.model.db.PadLockEntryModel
import com.squareup.sqlbrite2.BriteDatabase
import com.squareup.sqlbrite2.SqlBrite
import io.reactivex.Completable
import io.reactivex.Scheduler
import io.reactivex.Single
import timber.log.Timber
import javax.inject.Inject
import javax.inject.Named
import javax.inject.Singleton

@Singleton internal class PadLockDBImpl @Inject internal constructor(context: Context,
        @param:Named("io") private val scheduler: Scheduler) : PadLockDBInsert,
        PadLockDBUpdate,
        PadLockDBQuery, PadLockDBDelete {

    private var briteDatabase: BriteDatabase? = null
    private val openHelper: PadLockOpenHelper

    init {
        openHelper = PadLockOpenHelper(context)
    }

    @CheckResult private fun openDatabase(): BriteDatabase {
        if (briteDatabase == null) {
            briteDatabase = SqlBrite.Builder().build().wrapDatabaseHelper(openHelper, scheduler)
        }

        return briteDatabase!!
    }

    @CheckResult private fun deleteWithPackageActivityNameUnguarded(packageName: String,
            activityName: String): Int =
            PadLockEntry.deletePackageActivity(openHelper).executeProgram(packageName, activityName)

    @CheckResult override fun insert(packageName: String, activityName: String,
            lockCode: String?, lockUntilTime: Long, ignoreUntilTime: Long, isSystem: Boolean,
            whitelist: Boolean): Completable {
        return Single.fromCallable {
            val entry = PadLockEntry.create(packageName, activityName, lockCode, lockUntilTime,
                    ignoreUntilTime,
                    isSystem, whitelist)
            if (PadLockEntry.isEmpty(entry)) {
                throw RuntimeException("Cannot insert EMPTY entry")
            }
            Timber.i("DB: INSERT")
            val deleteResult = deleteWithPackageActivityNameUnguarded(packageName, activityName)
            Timber.d("Delete result: %d", deleteResult)
            return@fromCallable entry
        }.flatMapCompletable {
            Completable.fromAction { PadLockEntry.insertEntry(openHelper).executeProgram(it) }
        }
    }

    override fun updateIgnoreTime(ignoreUntilTime: Long, packageName: String,
            activityName: String): Completable {
        return Completable.fromCallable {
            if (PadLockEntry.PACKAGE_EMPTY == packageName || PadLockEntry.ACTIVITY_EMPTY == activityName) {
                throw RuntimeException("Cannot update EMPTY entry")
            }
            Timber.i("DB: UPDATE IGNORE")
            return@fromCallable PadLockEntry.updateIgnoreTime(openHelper)
                    .executeProgram(ignoreUntilTime, packageName, activityName)
        }
    }

    override fun updateLockTime(lockUntilTime: Long, packageName: String,
            activityName: String): Completable {
        return Completable.fromCallable {
            if (PadLockEntry.PACKAGE_EMPTY == packageName || PadLockEntry.ACTIVITY_EMPTY == activityName) {
                throw RuntimeException("Cannot update EMPTY entry")
            }
            Timber.i("DB: UPDATE LOCK")
            return@fromCallable PadLockEntry.updateLockTime(openHelper)
                    .executeProgram(lockUntilTime, packageName, activityName)
        }
    }

    override fun updateWhitelist(whitelist: Boolean, packageName: String,
            activityName: String): Completable {
        return Completable.fromCallable {
            if (PadLockEntry.PACKAGE_EMPTY == packageName || PadLockEntry.ACTIVITY_EMPTY == activityName) {
                throw RuntimeException("Cannot update EMPTY entry")
            }
            Timber.i("DB: UPDATE WHITELIST")
            return@fromCallable PadLockEntry.updateWhitelist(openHelper)
                    .executeProgram(whitelist, packageName, activityName)
        }
    }

    /**
     * Get either the package with specific name of the PACKAGE entry

     * SQLite only has bindings so we must make do
     * ?1 package name
     * ?2 the PadLock PACKAGE_TAG, see model.PadLockEntry
     * ?3 the specific activity name
     * ?4 the PadLock PACKAGE_TAG, see model.PadLockEntry
     * ?5 the specific activity name
     */
    @CheckResult override fun queryWithPackageActivityNameDefault(packageName: String,
            activityName: String): Single<PadLockEntry> {
        return Single.defer {
            Timber.i("DB: QUERY PACKAGE ACTIVITY DEFAULT")
            val statement = PadLockEntry.withPackageActivityNameDefault(packageName, activityName)
            return@defer openDatabase().createQuery(statement.tables, statement.statement,
                    *statement.args)
                    .mapToOne { PadLockEntry.WITH_PACKAGE_ACTIVITY_NAME_DEFAULT_MAPPER.map(it) }
                    .first(PadLockEntry.EMPTY)
        }
    }

    @CheckResult override fun queryWithPackageName(
            packageName: String): Single<List<PadLockEntry.WithPackageName>> {
        return Single.defer {
            Timber.i("DB: QUERY PACKAGE")
            val statement = PadLockEntry.withPackageName(packageName)
            return@defer openDatabase().createQuery(statement.tables, statement.statement,
                    *statement.args)
                    .mapToList { PadLockEntry.WITH_PACKAGE_NAME_MAPPER.map(it) }
                    .first(emptyList())
        }
    }

    @CheckResult override fun queryAll(): Single<List<PadLockEntry.AllEntries>> {
        return Single.defer {
            Timber.i("DB: QUERY ALL")
            val statement = PadLockEntry.queryAll()
            return@defer openDatabase().createQuery(statement.tables, statement.statement,
                    *statement.args)
                    .mapToList { PadLockEntry.ALL_ENTRIES_MAPPER.map(it) }
                    .first(emptyList())
        }
    }

    @CheckResult override fun deleteWithPackageName(packageName: String): Completable {
        return Completable.fromCallable {
            Timber.i("DB: DELETE PACKAGE")
            return@fromCallable PadLockEntry.deletePackage(openHelper).executeProgram(packageName)
        }
    }

    @CheckResult override fun deleteWithPackageActivityName(packageName: String,
            activityName: String): Completable {
        return Completable.fromCallable {
            Timber.i("DB: DELETE PACKAGE ACTIVITY")
            return@fromCallable deleteWithPackageActivityNameUnguarded(packageName, activityName)
        }
    }

    @CheckResult override fun deleteAll(): Completable {
        return Completable.fromAction {
            Timber.i("DB: DELETE ALL")
            openDatabase().use {
                it.execute(PadLockEntryModel.DELETE_ALL)
            }
        }.andThen(Completable.fromAction { openHelper.deleteDatabase() })
                .andThen(Completable.fromAction { briteDatabase = null })
                .andThen(Completable.fromAction { PadLockEntry.reset() })
    }

    private class PadLockOpenHelper internal constructor(context: Context) : SQLiteOpenHelper(
            context.applicationContext,
            DB_NAME, null,
            DATABASE_VERSION) {

        private val appContext: Context = context.applicationContext

        internal fun deleteDatabase() {
            appContext.deleteDatabase(
                    DB_NAME)
        }

        override fun onCreate(sqLiteDatabase: SQLiteDatabase) {
            Timber.d("onCreate")
            Timber.d("EXEC SQL: %s", PadLockEntryModel.CREATE_TABLE)
            sqLiteDatabase.execSQL(PadLockEntryModel.CREATE_TABLE)
        }

        override fun onUpgrade(sqLiteDatabase: SQLiteDatabase, oldVersion: Int, newVersion: Int) {
            Timber.d("onUpgrade from old version %d to new %d", oldVersion, newVersion)
            var currentVersion = oldVersion
            if (currentVersion == 1 && newVersion >= 2) {
                upgradeVersion1To2(sqLiteDatabase)
                ++currentVersion
            }

            if (currentVersion == 2 && newVersion >= 3) {
                upgradeVersion2To3(sqLiteDatabase)
                ++currentVersion
            }

            if (currentVersion == 3 && newVersion >= 4) {
                upgradeVersion3To4(sqLiteDatabase)
                ++currentVersion
            }
        }

        private fun upgradeVersion3To4(sqLiteDatabase: SQLiteDatabase) {
            Timber.d("Upgrading from Version 2 to 3 adds whitelist column")
            val alterWithWhitelist = "ALTER TABLE ${PadLockEntryModel.TABLE_NAME} ADD COLUMN ${PadLockEntryModel.WHITELIST} INTEGER NOT NULL DEFAULT 0"
            Timber.d("EXEC SQL: %s", alterWithWhitelist)
            sqLiteDatabase.execSQL(alterWithWhitelist)
        }

        private fun upgradeVersion2To3(sqLiteDatabase: SQLiteDatabase) {
            Timber.d("Upgrading from Version 2 to 3 drops the whole table")

            val dropOldTable = "DROP TABLE ${PadLockEntryModel.TABLE_NAME}"
            Timber.d("EXEC SQL: %s", dropOldTable)
            sqLiteDatabase.execSQL(dropOldTable)

            // Creating the table again
            onCreate(sqLiteDatabase)
        }

        private fun upgradeVersion1To2(sqLiteDatabase: SQLiteDatabase) {
            Timber.d("Upgrading from Version 1 to 2 drops the displayName column")

            // Remove the columns we don't want anymore from the table's list of columns
            Timber.d("Gather a list of the remaining columns")
            val columnsSeparated = TextUtils.join(",",
                    UPGRADE_1_TO_2_TABLE_COLUMNS)
            Timber.d("Column separated: %s", columnsSeparated)

            val tableName = PadLockEntryModel.TABLE_NAME
            val oldTable = tableName + "_old"
            val alterTable = "ALTER TABLE $tableName RENAME TO $oldTable"
            val insertIntoNewTable = "INSERT INTO $tableName($columnsSeparated) SELECT $columnsSeparated FROM $oldTable"
            val dropOldTable = "DROP TABLE $oldTable"

            // Move the existing table to an old table
            Timber.d("EXEC SQL: %s", alterTable)
            sqLiteDatabase.execSQL(alterTable)

            onCreate(sqLiteDatabase)

            // Populating the table with the data
            Timber.d("EXEC SQL: %s", insertIntoNewTable)
            sqLiteDatabase.execSQL(insertIntoNewTable)

            Timber.d("EXEC SQL: %s", dropOldTable)
            sqLiteDatabase.execSQL(dropOldTable)
        }

        companion object {

            private const val DB_NAME = "padlock_db"
            private const val DATABASE_VERSION = 4

            private val UPGRADE_1_TO_2_TABLE_COLUMNS = arrayOf(PadLockEntryModel.PACKAGENAME,
                    PadLockEntryModel.ACTIVITYNAME, PadLockEntryModel.LOCKCODE,
                    PadLockEntryModel.LOCKUNTILTIME,
                    PadLockEntryModel.IGNOREUNTILTIME, PadLockEntryModel.SYSTEMAPPLICATION)
        }
    }
}